### DISCUSSION TITLE
Newbie feedback

### DISCUSSION THREAD

**Discussion by clydebarrow** [scottlamb/moonfire-nvr#151] [Q&A]
I've just started setting up Moonfire - I plan to have two installs, one to replace my Swann/Hikvision NVR at home, and a new one in an RV where power consumption is critical. I tried Shinobi and gave up because of the byzantine user interface.

I have some feedback and a few questions from the install and setup process. This is not yet complete but I want to get this stuff down as I encounter it, as it is intended to be constructive.

### My system:

* Raspberry Pi 4 (8GB memory)
* Storage: 32GB Sandisk SD card and a Sandisk 1TB SSD.
* Cameras are a mix of Swann, Hikvision and Anpviz.
* OS is Ubuntu 64 bit server 20.04.2

### Problems I encountered and solved or worked around:

1. Docker - it seems I need to run the nvr script (i.e. nvr config, nvr run) as root. I installed the Docker rootless config but that didn't work.
2. RTSP Url setup - the config asks for the host, user and password and a url. I initially entered the url with username/password but when starting the server (nvr run) I got no open port on 8080. After figuring out how to see the logs I got a long stack trace that started with an exception complaining that the url should not include the username/password. So I changed the url to just the path, i.e. `/Streaming/Channels/101`, still no joy, now complaining about a relative url. I finally figured out that it needs the url *with* the scheme and host, but *not* the username/password. This should be picked up when entering it in the config, not with an exception at run time. And...
3. It seems I have to stop the container to run `nvr config`. This, plus the need to run `nvr logs` to actually see what is going on, makes the setup process for a newbie like me somewhat frustrating.
4. Similarly to 3, if I don't set a sample directory the config dialog is quite happy but an exception is thrown at run time.
5. The logs out of Docker have a full stacktrace for an exception, which seems superfluous and makes reading them hard.

### Current problems

I have set up one camera (Hikvision) that works fine. Another camera (Anpviz IPC-D250) works fine with VLC, my Hikvision NVR, and Shinobi, but not Moonfire - it passes the Test in Config (the stream resolution is correctly identified) but will not record. There are two problems:

1. (worked-around) If I set up username and password in the config I see these errors:
````
W20210823 05:40:40.871 s-ChuckRear-sub moonfire_nvr::streamer] ChuckRear-sub: sleeping for Duration { secs: 1, nanos: 0 } after error: [192.168.254.254:39730(me)->192.168.254.5:554@2021-08-23T05:40:40, 0@2021-08-23T05:40:40] Unauthorized response to DESCRIBE CSeq=1: Non-digest authentication requested: Basic realm="/"
````

2. If I leave the username and password blank and add them as parameters to the RTSP Url then I see exceptions like this:
````
I20210823 05:44:11.893 s-ChuckRear-sub moonfire_nvr::streamer] ChuckRear-sub: Opening input: rtsp://192.168.254.5/h264cif?user=User&password=xxxx
W20210823 05:44:12.039 s-ChuckRear-sub moonfire_nvr::streamer] ChuckRear-sub: sleeping for Duration { secs: 1, nanos: 0 } after error: [192.168.254.254:39968(me)->192.168.254.5:554@2021-08-23T05:44:11, 0@2021-08-23T05:44:12] Ok response to DESCRIBE CSeq=1: Unable to parse SDP: SdpInvalidSyntax: a=
Length: 819 (0x333) bytes
0000:   76 3d 30 0d  0a 6f 3d 2d  20 31 31 30  39 31 36 32   v=0..o=- 1109162
0010:   30 31 34 32  31 39 31 38  32 20 31 31  30 39 31 36   014219182 110916
0020:   32 30 31 34  32 31 39 31  39 32 20 49  4e 20 49 50   2014219192 IN IP
0030:   34 20 78 2e  79 2e 7a 2e  77 0d 0a 73  3d 52 54 53   4 x.y.z.w..s=RTS
0040:   50 2f 52 54  50 20 73 74  72 65 61 6d  20 66 72 6f   P/RTP stream fro
0050:   6d 20 69 63  61 6d 72 61  20 69 70 63  0d 0a 65 3d   m icamra ipc..e=
...
````

This looks to me like the camera is not standards-compliant, but it does work fine with lots of other things.

3. The times of the recordings are wrong. It's not my local time and it's not UTC. Where is the timezone set?

More to come.

**Comment by scottlamb** [comment]
Thanks for writing this up!

A few of these things are because Moonfire is in the process of switching its RTSP support from ffmpeg (which despite its problems has a lot more flight miles) to Retina (a new library I've written myself). Long-term I want to get Retina working well, but for now it's possible to switch back to ffmpeg via `--rtsp-library=ffmpeg`. You'd add this to the `docker run` command in `/usr/local/bin/nvr` next to the `--allowunauthenticated-permissions=...` argument and use `nvr stop && nvr rm && nvr run` to make it take effect.

> 1. Docker - it seems I need to run the nvr script (i.e. nvr config, nvr run) as root. I installed the Docker rootless config but that didn't work.

You're the second person this week to bring this up. #142 I'll definitely adjust the instructions a bit.

> 2. RTSP Url Setup - ... After figuring out how to see the logs I got a long stack trace that started with an exception complaining that the url should not include the username/password

Good point. Filed #152.

> 3. It seems I have to stop the container to run nvr config. This, plus the need to run nvr logs to actually see what is going on, makes the setup process for a newbie like me somewhat frustrating.

Yeah, that's how it is for the moment. Eventually I'd like to have the configuration just be part of the web UI (#35), but it's a lot of work.

> 4. Similarly to 3, if I don't set a sample directory the config dialog is quite happy but an exception is thrown at run time.

#152 also.

> Unauthorized response to DESCRIBE CSeq=1: Non-digest authentication requested: Basic realm="/"

That'll be easy to support. Filed <https://github.com/scottlamb/retina/issues/25>.

> Unable to parse SDP: SdpInvalidSyntax: a=
Length: 819 (0x333) bytes
0000:   76 3d 30 0d  0a 6f 3d 2d  20 31 31 30  39 31 36 32   v=0..o=- 1109162
0010:   30 31 34 32  31 39 31 38  32 20 31 31  30 39 31 36   014219182 110916
0020:   32 30 31 34  32 31 39 31  39 32 20 49  4e 20 49 50   2014219192 IN IP
0030:   34 20 78 2e  79 2e 7a 2e  77 0d 0a 73  3d 52 54 53   4 x.y.z.w..s=RTS
0040:   50 2f 52 54  50 20 73 74  72 65 61 6d  20 66 72 6f   P/RTP stream fro
0050:   6d 20 69 63  61 6d 72 61  20 69 70 63  0d 0a 65 3d   m icamra ipc..e=
...

Can you paste the remaining lines of this message? We can probably get this one fixed too then.

> The times of the recordings are wrong. It's not my local time and it's not UTC. Where is the timezone set?

The first line of your `/usr/local/bin/nvr` wrapper script.

**Comment by clydebarrow** [comment]
I can confirm that using ffmpeg gets the camera working. 

Here's the full dump of one of the examples of unparsed data:

````
Length: 827 (0x33b) bytes
0000:   76 3d 30 0d  0a 6f 3d 2d  20 31 31 30  39 31 36 32   v=0..o=- 1109162
0010:   30 31 34 32  31 39 31 38  32 20 31 31  30 39 31 36   014219182 110916
0020:   32 30 31 34  32 31 39 31  39 32 20 49  4e 20 49 50   2014219192 IN IP
0030:   34 20 78 2e  79 2e 7a 2e  77 0d 0a 73  3d 52 54 53   4 x.y.z.w..s=RTS
0040:   50 2f 52 54  50 20 73 74  72 65 61 6d  20 66 72 6f   P/RTP stream fro
0050:   6d 20 61 6e  6a 76 69 73  69 6f 6e 20  69 70 63 61   m anjvision ipca
0060:   6d 65 72 61  0d 0a 65 3d  4e 4f 4e 45  0d 0a 63 3d   mera..e=NONE..c=
0070:   49 4e 20 49  50 34 20 30  2e 30 2e 30  2e 30 0d 0a   IN IP4 0.0.0.0..
0080:   61 3d 74 6f  6f 6c 3a 4c  49 56 45 35  35 35 20 53   a=tool:LIVE555 S
0090:   74 72 65 61  6d 69 6e 67  20 4d 65 64  69 61 20 76   treaming Media v
00a0:   32 30 31 31  2e 30 35 2e  32 35 20 43  48 41 4d 2e   2011.05.25 CHAM.
00b0:   4c 49 40 41  4e 4a 56 49  53 49 4f 4e  2e 43 4f 4d   LI@ANJVISION.COM
00c0:   0d 0a 74 3d  30 20 30 0d  0a 61 3d 72  61 6e 67 65   ..t=0 0..a=range
00d0:   3a 6e 70 74  3d 30 2d 0d  0a 61 3d 63  6f 6e 74 72   :npt=0-..a=contr
00e0:   6f 6c 3a 2a  0d 0a 6d 3d  76 69 64 65  6f 20 30 20   ol:*..m=video 0
00f0:   52 54 50 2f  41 56 50 20  39 36 0d 0a  61 3d 72 74   RTP/AVP 96..a=rt
0100:   70 6d 61 70  3a 39 36 20  48 32 36 34  2f 39 30 30   pmap:96 H264/900
0110:   30 30 0d 0a  61 3d 63 6f  6e 74 72 6f  6c 3a 74 72   00..a=control:tr
0120:   61 63 6b 49  44 3d 31 0d  0a 61 3d 66  6d 74 70 3a   ackID=1..a=fmtp:
0130:   39 36 20 70  72 6f 66 69  6c 65 2d 6c  65 76 65 6c   96 profile-level
0140:   2d 69 64 3d  34 44 34 30  31 46 3b 70  61 63 6b 65   -id=4D401F;packe
0150:   74 69 7a 61  74 69 6f 6e  2d 6d 6f 64  65 3d 30 3b   tization-mode=0;
0160:   73 70 72 6f  70 2d 70 61  72 61 6d 65  74 65 72 2d   sprop-parameter-
0170:   73 65 74 73  3d 5a 30 31  41 48 35 57  67 4c 41 53   sets=Z01AH5WgLAS
0180:   61 62 41 51  3d 2c 61 4f  34 38 67 41  3d 3d 3b 63   abAQ=,aO48gA==;c
0190:   6f 6e 66 69  67 3d 30 30  30 30 30 30  30 31 36 37   onfig=0000000167
01a0:   34 64 34 30  31 66 39 35  61 30 32 63  30 34 39 61   4d401f95a02c049a
01b0:   36 63 30 34  30 30 30 30  30 30 30 31  36 38 65 65   6c040000000168ee
01c0:   33 63 38 30  30 30 30 30  30 30 30 31  30 36 66 30   3c800000000106f0
01d0:   32 63 30 34  34 35 63 36  66 35 30 30  30 36 32 30   2c0445c6f5000620
01e0:   65 62 63 32  66 33 66 37  36 33 39 65  34 38 32 35   ebc2f3f7639e4825
01f0:   30 62 66 63  62 35 36 31  62 62 32 62  38 35 64 64   0bfcb561bb2b85dd
0200:   61 36 66 65  35 66 30 36  63 63 38 62  38 38 37 62   a6fe5f06cc8b887b
0210:   36 61 39 31  35 66 35 61  61 33 62 65  62 66 66 66   6a915f5aa3bebfff
0220:   66 66 66 66  66 66 66 66  37 33 38 30  0d 0a 61 3d   ffffffff7380..a=
0230:   78 2d 64 69  6d 65 6e 73  69 6f 6e 73  3a 20 37 30   x-dimensions: 70
0240:   34 2c 20 35  37 36 0d 0a  61 3d 78 2d  66 72 61 6d   4, 576..a=x-fram
0250:   65 72 61 74  65 3a 20 31  32 0d 0a 6d  3d 61 75 64   erate: 12..m=aud
0260:   69 6f 20 30  20 52 54 50  2f 41 56 50  20 30 0d 0a   io 0 RTP/AVP 0..
0270:   61 3d 72 74  70 6d 61 70  3a 30 20 4d  50 45 47 34   a=rtpmap:0 MPEG4
0280:   2d 47 45 4e  45 52 49 43  2f 31 36 30  30 30 2f 32   -GENERIC/16000/2
0290:   0d 0a 61 3d  66 6d 74 70  3a 30 20 63  6f 6e 66 69   ..a=fmtp:0 confi
02a0:   67 3d 31 34  30 38 0d 0a  61 3d 63 6f  6e 74 72 6f   g=1408..a=contro
02b0:   6c 3a 74 72  61 63 6b 49  44 3d 32 0d  0a 61 3d 4d   l:trackID=2..a=M
02c0:   65 64 69 61  5f 68 65 61  64 65 72 3a  4d 45 44 49   edia_header:MEDI
02d0:   41 49 4e 46  4f 3d 34 39  34 44 34 42  34 38 30 31   AINFO=494D4B4801
02e0:   30 31 30 30  30 30 30 34  30 30 30 31  30 30 31 30   0100000400010010
02f0:   37 31 30 31  31 30 34 30  31 46 30 30  30 30 30 30   710110401F000000
0300:   46 41 30 30  30 30 30 30  30 30 30 30  30 30 30 30   FA00000000000000
0310:   30 30 30 30  30 30 30 30  30 30 30 30  30 30 30 30   0000000000000000
0320:   30 30 30 30  30 30 3b 0d  0a 61 3d 61  70 70 76 65   000000;..a=appve
0330:   72 73 69 6f  6e 3a 31 2e  30 0d 0a                   rsion:1.0..
````

WRT to the timezone, I changed the first line of the nvr script to:

````
tz=`cat /etc/timezone`
````

So at this point it's working pretty well with 2 cameras. I've got more to do, but am pretty pleased. CPU usage is very low. I have other things to set up and questions related to them but will read the docs first as I progress.

I have one further issue - sometimes when trying to play back a recording I get a message "No video with supported format and MIME type found." Other times it works fine on the same recording and the live view always works.

<img width="423" alt="Screenshot 2021-08-24 at 8 16 20 am" src="https://user-images.githubusercontent.com/2366188/130526468-4c47aba7-0b4b-4fe5-8903-29eb7a4fa32b.png">

Other UI tweaks I'd like to see:

1. When playing back a recording I can stop and start the playback by clicking, and close the window by ESC, but only after I have clicked on the window (Hmm, maybe the overlay window is just not being given focus when opened.) I expected ESC to always close the window, and spacebar to toggle playback.
2. The live view always reverts to nothing playing when opened. I'd like it to restore the last used layout and streams. Configurable profiles would be nice too :-)

Again, this is all intended as constructive feedback - I really appreciate what you've done here and think it's going to suit what I need very well. You can expect to see PRs from me further down the track.

**Reply by scottlamb** [reply]
Thanks. That dump helped me understand the SDP problem. Filed scottlamb/retina#26.

**Reply by scottlamb** [reply]
Could you please file a bug issue about the "No video with supported format and MIME type found" error? I've never seen it before and will need some more info to understand. The bug template will ask for some stuff to start, like browser version.

Those UI tweaks are things I'd like too. I'm not an expert in UIs so this stuff takes me quite a while. Help welcome!

**Reply by clydebarrow** [reply]
> Could you please file a bug issue about the "No video with supported format and MIME type found" error? 

I can't reproduce it now. I suspect it was due to the fact that I had not set any value for the disk space allocated for each stream, so the stream files were being deleted as fast as they were created. The disk space setting should probably be in the stream config, not the directory config, and have a default non-zero value to stop people like me shooting themselves in the foot.

**Reply by scottlamb** [reply]
That's not the error I'd expect to see if the files are deleted, but I suppose it's the browser that decides how to report the error for now. (What it'd see is either a 404 status or a truncated HTTP response, depending on when the deletion happens.)

It's in the directory config because you typically want to adjust them all at once and look at the totals vs the filesystem capacity.

Some day we'll have a web UI that's really intuitive, but TUIs are a bit constraining.

**Reply by scottlamb** [reply]
FYI, besides the other two problems you've encountered with this Anpviz camera, I can see from the `a=tool:LIVE555 Streaming Media v2011.05.25 CHAM.LI@ANJVISION.COM` that you're going to hit https://github.com/scottlamb/retina/issues/17

**Comment by clydebarrow** [comment]
Some more questions:

1. The install guide says *Be sure to assign each stream you want to capture to a sample file directory and check the "record" box.*  
Can a single directory be used for all streams, or does each stream need a unique directory?     

2. *Assign disk space to your cameras back in "Directories and retention". Leave a little slack between the total limit and the filesystem capacity, even if you store nothing else on the disk. 1 GiB per camera should be plenty.*
I assume the 1GiB refers to the slack needed, not the disk space assigned?

**Reply by scottlamb** [reply]
> Can a single directory be used for all streams, or does each stream need a unique directory?

One for all streams is fine. I use one per hard drive.

> I assume the 1GiB refers to the slack needed, not the disk space assigned?

Yes.

**Reply by scottlamb** [reply]
7ed02bd should answer these questions in the guide for the next person.
